### DISCUSSION TITLE
CTT some failed tests Solved

### DISCUSSION THREAD

**Discussion by johannwesely** [eclipse-milo/milo#1288] [General]
Hi,
I think i solved some problems regarding ctt and want to ask if my solutions seems legit and if you could add them to dev1.0?
I had an Issue with CTT Script Base Information->Base Info Type System-> ReferenceTypesValidation 
![CTT ReferenceTypes 2024-07-30 145550](https://github.com/user-attachments/assets/c087cadd-3158-46f5-9675-07378f983106)

I added in AttributeReader Line 156 a symmetric check as mentioned in part 3,5.2

 
```
switch (attributeId) {
                case DataTypeDefinition:
                case RolePermissions:
                case UserRolePermissions:
                case AccessRestrictions:
                case AccessLevelEx:
                    if (value == null) {
                        return new DataValue(StatusCodes.Bad_AttributeIdInvalid);
                    }

            }
            if (attributeId == AttributeId.InverseName && node instanceof UaReferenceTypeNode referenceNode){
                if( referenceNode.getSymmetric()){
                    return new DataValue(StatusCodes.Bad_AttributeIdInvalid);
                }
            }
```

![refrencetype green 2024-07-30 150114](https://github.com/user-attachments/assets/45ac62ff-2380-4133-87d9-a373d2da6e42)
and im green

Also I had Problem with ServerDiagnostics 
![Diagnostics 2024-07-30 150423](https://github.com/user-attachments/assets/7b748afc-5452-4e18-a312-f144d2ea960b)
here i set in OPcUaNamespace the ServerDiagnostic Enabled Flag in line 186 ` serverTypeNode.getServerDiagnosticsNode().setEnabledFlag(true);` 
now i have an new error(ConditionType not found) but Timestamps are correct now.

How is this flag meant to be set must the client first enable and then use diagnostics ? But Ctt doesn't do this in script.

Im Trying to figure out more CTT failure solutions, but i dont know if the Forum is the right place to post them?

Kind Regards 
Johann Wesely

**Comment by kevinherron** [comment]
I'll try to look into these today.

> Im Trying to figure out more CTT failure solutions, but i dont know if the Forum is the right place to post them?

This is fine, but in my experience with the CTT in the past it's also likely you'll find bugs in the CTT scripts, which need to be reported to the OPC Foundation Mantis issue tracker.

What version of the CTT are you using?

**Reply by johannwesely** [reply]
I Send you the About Screen because i dont know which is the relevant version for you
![ctt 2024-07-29 112001](https://github.com/user-attachments/assets/e8d653c6-77d7-439b-b608-2f67c819d9de)

**Comment by kevinherron** [comment]
Notes to self:
- `inverseName` field in `UaReferenceTypeNode`  and parameter in constructor should be marked `@Nullable`
- switch starting on `AttributeReader:143` needs to include `AttributeId.InverseName`
- `ReferenceTypeNodeLoader` needs to construct using `null` values instead of `LocalizedText.NULL_VALUE` for `inverseName`

**Comment by kevinherron** [comment]
@johannwesely I don't have a CTT set up right now. Please test https://github.com/eclipse/milo/pull/1292 when you have a chance.

**Reply by johannwesely** [reply]
Hi Kevin, i tested gh1291 it seems that some of the Symmetric ReferenceTypes failures are Good now but 2 failures still exist: 
![image](https://github.com/user-attachments/assets/16b090bf-cd2e-4692-9fa2-f4e33ccaf5f7)
right now i have no time to investigate further but maybe later this day.

**Reply by johannwesely** [reply]
The Failing id's 6030 is ConnectsTo and 6467 is connectstoParent maybe this helps.

**Reply by johannwesely** [reply]
Hi, i had the time to investigate this. I think the Problem is that SIOMEV2.7.1 always has an InverseName Set to symmetric references as standard. When I parse the .xml from Siome to milo Namespace i have InvereName set and not null so i think your fix should be correct. 
![image](https://github.com/user-attachments/assets/fce9111a-4e85-45a3-979e-0f9b060eb697)
